temporalio/contrib/opentelemetry.py

```diff
@@ -0,0 +1,544 @@
+import opentelemetry.util.types
+from typing_extensions import Protocol, TypeAlias, TypedDict
+
+import temporalio.activity
+import temporalio.api.common.v1
+import temporalio.client
+import temporalio.converter
+import temporalio.worker
+import temporalio.workflow
+
+default_text_map_propagator = opentelemetry.propagators.composite.CompositePropagator(
+    [
+        opentelemetry.trace.propagation.tracecontext.TraceContextTextMapPropagator(),
+        opentelemetry.baggage.propagation.W3CBaggagePropagator(),
+    ]
+)
+"""Default text map propagator used by :py:class:`TracingInterceptor`."""
+
+_CarrierDict: TypeAlias = Dict[str, opentelemetry.propagators.textmap.CarrierValT]
+
+
+class TracingInterceptor(temporalio.client.Interceptor, temporalio.worker.Interceptor):
+    """Interceptor that supports client and worker OpenTelemetry span creation
+    and propagation.
+
+    This should be created and used for ``interceptors`` on the
+    :py:meth:`temporalio.client.Client.connect` call to apply to all client
+    calls and worker calls using that client. To only apply to workers, set as
+    worker creation option instead of in client.
+
+    To customize the header key, text map propagator, or payload converter, a
+    subclass of this and :py:class:`TracingWorkflowInboundInterceptor` should be
+    created. In addition to customizing those attributes, the subclass of this
+    class should return the workflow interceptor subclass from
+    :py:meth:`workflow_interceptor_class`. That subclass should also set the
+    custom attributes desired.
+    """
+
+    def __init__(
+        self,
+        tracer: Optional[opentelemetry.trace.Tracer] = None,
+    ) -> None:
+        """Initialize a OpenTelemetry tracing interceptor.
+
+        Args:
+            tracer: The tracer to use. Defaults to
+                :py:func:`opentelemetry.trace.get_tracer`. Currently, the tracer
+                must be an instance of
+                :py:class:`opentelemetry.sdk.trace.Tracer` due to internal
+                serialization approaches used.
+        """
+        self.tracer = tracer or opentelemetry.trace.get_tracer(__name__)
+        # Due to the fact that we have to violate the API to make spans
+        # writeable after deserializing so it can cross the sandbox, we must
+        # have an SDK tracer instance
+        if not isinstance(self.tracer, opentelemetry.sdk.trace.Tracer):
+            raise ValueError("Currently only SDK tracers supported")
+        # To customize any of this, users must subclass. We intentionally don't
+        # accept this in the constructor because if they're customizing these
+        # values, they'd also need to do it on the workflow side via subclassing
+        # on that interceptor since they can't accept custom constructor values.
+        self.header_key: str = "_tracer-data"
+        self.text_map_propagator: opentelemetry.propagators.textmap.TextMapPropagator = (
+            default_text_map_propagator
```

Comment:
In TS we use the global propagator so users can decide whether they want to propagate baggage or use the jaeger propagator (default when using opentracing).

https://docs.temporal.io/typescript/logging#context-propagation

Reply:
Ah, I didn't realize there was a global, but I see it now at https://opentelemetry-python.readthedocs.io/en/latest/api/propagate.html#opentelemetry.propagate.get_global_textmap. I will switch to that (but still keep it a class member so the class can be extended to customize it).